build/build.go

```diff
@@ -591,24 +611,64 @@ func (s *Session) buildImportPathWithSrcDir(path string, srcDir string) (*Packag
 	return pkg, archive, nil
 }
 
+func hashCompilerBinary() (string, error) {
+	if compilerBinaryHash != "" {
+		return compilerBinaryHash, nil
+	}
+
+	binHash := sha256.New()
+	binPath, err := os.Executable()
+	if err != nil {
+		return "", fmt.Errorf("could not locate GopherJS binary: %v", err)
+	}
+	binFile, err := os.Open(binPath)
+	if err != nil {
+		return "", fmt.Errorf("could not open %v: %v", binPath, err)
+	}
+	defer binFile.Close()
+	if _, err := io.Copy(binHash, binFile); err != nil {
+		return "", fmt.Errorf("failed to hash %v: %v", binPath, err)
+	}
+	compilerBinaryHash = fmt.Sprintf("%#x", binHash.Sum(nil))
+	return compilerBinaryHash, nil
+}
+
 func (s *Session) BuildPackage(pkg *PackageData) (*compiler.Archive, error) {
 	if archive, ok := s.Archives[pkg.ImportPath]; ok {
 		return archive, nil
 	}
 
+	// For non-main and test packages we build up a hash that will help
+	// determine staleness. Set hashDebug to see this in action. The format is:
+	//
+	// ## <package>
+	// compiler binary hash: 0x519d22c6ab65a950f5b6278e4d65cb75dbd3a7eb1cf16e976a40b9f1febc0446
+	// build tags: <list of build tags>
+	// import: <import path>
+	//   hash: 0xb966d7680c1c8ca75026f993c153aff0102dc9551f314e5352043187b5f9c9a6
+	// ...
+	//
+	// file: <file path>
+	// <file contents>
+	// N bytes
```

Comment:
Hmm, I feel like this format is a little bit arbitrary. Can we use JSON or YAML or something formatted? I don't have a strong opinion though...

Reply:
A possible problem with JSON/Yaml is that it's not inherently ordered. While we could find a way to ensure constant ordering, it wouldn't necessarily happen by default, which could lead to false negative cache hits.

Reply:
The goal here is to create a hash as quickly as possible in order to determine whether we have a cache miss or not. So yes, whilst the format is arbitrary, it is simple. And requires no additional processing in order to write to the hash. Using JSON/YAML adds more overhead in the middle, overhead that is unnecessary because ultimately the output is a 256 bit value. Humans will only ever read this whilst debugging (which itself will be a rare occurrence) with `hashDebug = true`.